---
layout: post
title: "PhD thesis: Probabilistic Modelling of Genomic Trajectories"
---

In September I successfully defended my DPhil at Oxford on "Probabilistic Modelling of Genomic Trajectories". This tackles the single-cell pseudotime problem from a (mostly Bayesian) statistical latent variable approach. 

The [**entire thesis**]({{site.baseurl}}/files/krc-thesis.pdf) is available (39Mb) or broken down by chapter below. If you find any typos please let me know (but don't tell Oxford).

* [**Chapter 1**]({{site.baseurl}}/files/krc-thesis-chap-1.pdf) An introduction to single-cell RNA-seq analysis, the pseudotime estimation problem, and statistical latent variable modelling
* [**Chapter 2**]({{site.baseurl}}/files/krc-thesis-chap-2.pdf) Examining uncertainty in single-cell pseudotime using Gaussian Process Latent Variable Models
* [**Chapter 3**]({{site.baseurl}}/files/krc-thesis-chap-3.pdf) Learning single-cell pseudotime using Bayesian nonlinear factor analysis models
* [**Chapter 4**]({{site.baseurl}}/files/krc-thesis-chap-4.pdf) Inferring bifurcations using a Bayesian hierarchical mixture of factor analysers
* [**Chapter 5**]({{site.baseurl}}/files/krc-thesis-chap-5.pdf) introduces covariate-adjusted latent variable models for learning latent variables from heterogeneous samples (e.g. with different genetic/environmental backgrounds)